block/block_service.proto

```diff
@@ -71,23 +38,157 @@ import "stream/block.proto";
 import "stream/block_item.proto";
 
 /**
- * Publish a stream of blocks.
+ * Publish a stream of block items.
  *
- * Each item in the stream MUST contain one `BlockItem`.<br/>
+ * Each request in the stream MUST contain at least one `BlockItem`.<br/>
+ * Each request MAY contain more than one `BlockItem`.<br/>
+ * A single request MUST NOT contain `BlockItem`s from more than one block.<br/>
+ * Each request MAY contain a variable number of `BlockItem`s.<br/>
  * Each Block MUST begin with a single `BlockHeader` block item.<br/>
+ * If a `BlockHeader` is present in a request, it MUST be the first `BlockItem`
+ * in the `block_items` list.<br/>
  * The block node SHALL append each `BlockItem` to an internal structure
  * to reconstruct full blocks.<br/>
  * The block node MUST verify the block proof for each block before sending a
  * response message acknowledging that block.<br/>
- * Each Block MUST end with a single `BlockStateProof` block item.<br/>
- * The block node MUST verify the Block using the `BlockStateProof` to
- * ensure all data was received and processed correctly.<br/>
+ * Each Block MUST end with a single `BlockProof` block item.<br/>
+ * If a `BlockProof` is present in a request, it MUST be the last `BlockItem`
```

Comment:
Does this mean that only one full block can be sent in a request? I don't think that's what this means, but the verbiage implies a maximum of one `BlockProof`

Reply:
A request should almost never be a full block.  A single streamed request (the API is a stream of request messages) is 0 or more `BlockItem`s.  What we require here is that _if_ a `BlockProof` is present in the batch, it _must_ be the _last_ item.  This is so that we _never_ include more than one block in a single batch.  The Block Node has to process this data _without parsing_ to the maximum extent possible, so we require that no single batch of block items will _ever_ contain data from more than one block.
A block _must_ begin with a BlockHeader, and _must_ end with a BlockProof.  By requiring that a Block Header is _first_ in a batch, and a Block Proof is _last_ in a (usually different) batch, we ensure no single batch contains items from 2 or more blocks.